src/contract_deltas.jl

```diff
@@ -0,0 +1,160 @@
+must be the root of this set.
+"""
+function _delta_inds_disjointsets(deltas::Vector{<:ITensor}, rootinds::Vector{<:Index})
+  inds_list = map(t -> collect(inds(t)), deltas)
+  deltainds = collect(Set(vcat(inds_list...)))
+  ds = DisjointSets(deltainds)
+  for t in deltas
+    i1, i2 = inds(t)
+    if find_root!(ds, i1) in rootinds
+      _root_union!(ds, find_root!(ds, i1), find_root!(ds, i2))
+    else
+      _root_union!(ds, find_root!(ds, i2), find_root!(ds, i1))
+    end
+  end
+  return ds
+end
+
+"""
+Given an input tensor network `tn`, remove redundent delta tensors
+in `tn` and change inds accordingly to make the output `tn` represent
+the same tensor network but with less delta tensors.
+
+========
+Example:
+  julia> is = [Index(2, string(i)) for i in 1:6]
+  julia> a = ITensor(is[1], is[2])
+  julia> b = ITensor(is[2], is[3])
+  julia> delta1 = delta(is[3], is[4])
+  julia> delta2 = delta(is[5], is[6])
+  julia> tn = ITensorNetwork([a, b, delta1, delta2])
+  julia> ITensorNetworks._contract_deltas(tn)
+  ITensorNetwork{Int64} with 3 vertices:
+  3-element Vector{Int64}:
+   1
+   2
+   4
+
+  and 1 edge(s):
+  1 => 2
+
+  with vertex data:
+  3-element Dictionaries.Dictionary{Int64, Any}
+   1 │ ((dim=2|id=457|"1"), (dim=2|id=296|"2"))
+   2 │ ((dim=2|id=296|"2"), (dim=2|id=613|"4"))
+   4 │ ((dim=2|id=626|"6"), (dim=2|id=237|"5"))
+"""
+function _contract_deltas(tn::ITensorNetwork)
+  tn = copy(tn)
+  network = Vector{ITensor}(tn)
+  deltas = filter(t -> is_delta(t), network)
+  outinds = noncommoninds(network...)
+  ds = _delta_inds_disjointsets(deltas, outinds)
+  deltainds = [ds...]
+  sim_deltainds = [find_root!(ds, i) for i in deltainds]
+  # `rem_vertex!(tn, v)` changes `vertices(tn)` in place.
+  # We copy it here so that the enumeration won't be affected.
+  for v in copy(vertices(tn))
+    if !is_delta(tn[v])
+      tn[v] = replaceinds(tn[v], deltainds, sim_deltainds)
+      continue
+    end
+    i1, i2 = inds(tn[v])
+    root = find_root!(ds, i1)
+    @assert root === find_root!(ds, i2)
+    if i1 != root && i1 in outinds
+      tn[v] = delta(i1, root)
+    elseif i2 != root && i2 in outinds
+      tn[v] = delta(i2, root)
+    else
+      rem_vertex!(tn, v)
+    end
+  end
+  return tn
+end
+
+"""
+  TODO: do we want to make it a public function?
+"""
+function _noncommoninds(partition::DataGraph)
+  networks = [Vector{ITensor}(partition[v]) for v in vertices(partition)]
+  network = vcat(networks...)
+  return noncommoninds(network...)
+end
+
+"""
+Given an input `partition`, contract redundent delta tensors in `partition`
+without changing the tensor network value. `root` is the root of the
+dfs_tree that defines the leaves.
+Note: for each vertex `v` of `partition`, the number of non-delta tensors
+  in `partition[v]` will not be changed.
+Note: only delta tensors of non-leaf vertices will be contracted.
+"""
+function _contract_deltas(partition::DataGraph; root=1)
```

Comment:
So the idea here is that it does the same thing as `_contract_deltas(tn::ITensorNetwork)` but only within the partitions, i.e. it doesn't remove deltas that cross partitions?

If so, couldn't that be implemented by looping through the subgraphs and calling `_contract_deltas(tn_subgraph::ITensorNetwork)`?

Also, should this be `root=first(vertices(partition))` to handle more general vertex partition names?

Reply:
so the idea is that it will also contract delta cross partitions; except the partition defined on leaf vertices. The constraint on leaf vertices is to make sure that the noncommininds of the partition will still be adjacent to the leaves after removing deltas (otherwise the leaf partition could be empty and the noncommoninds then will be adjacent to inner partitions, which is not a good case to deal with, especially for my density matrix algorithm implementation)

And great idea for the root interface, will do that

Reply:
therefore, calling `_contract_deltas(partition[v])` for each `v` doesn't work since since it will not contract delta cross partitions

Reply:
I see, thanks. That seems a bit specialized for using the general function name `_contract_deltas`. What if we named this version `_contract_deltas_ignore_leaf_partitions` to make that clearer?

Also, it seems better to implement this by calling out to `_contract_deltas(tn::ITensorNetwork)`, for example by making an ITensorNetwork built by unioning the tensor networks of the partitions living on the nonleaf vertices `nonleaf_vertices`, calling `_contract_deltas(tn::ITensorNetwork)`, and then partitioning the resulting network.

Reply:
that sounds good, let me think about how to implement it

Reply:
Done, please let me know if you have any questions

Reply:
Thanks for the quick change, looks reasonable!